### PR TITLE
wip: start to attempt to migrate rendered templates out of data structure

### PR DESCRIPTION
**Pull request by kylebrandt**
Lots in these changes so need to rest and look at it with a fresh page of eyes.

Problem right now is that when expanding things on the dashboard, sometimes there are templates, and other times there are not.

The following seems not right:

```
$ redis-cli -p 6389 LRANGE allIncidents 0 -1 | wc -l
   71077

# kbrandt at grove in ~/src/bosun.org/cmd/bosun on git:rendTempStore x [16:03:15]
$ redis-cli -p 6389 KEYS incidentById:\* | wc -l
   64492
```

Also maybe something to do with not storing when emailbody == body? 

Anyways, need a break. The dashboard loads much faster though. Even with a handful of alerts the "Setup" phase in marshal groups goes down by a factor of 10 (30 vs 300ms)

### CODE REVIEW

**Comment by kylebrandt** [comment]
Or 290ms vs 4.5sec for a hundreds of alerts. So if the dashboard is flood this reduces the load time a lot. Also that several seconds that redis isn't being blocked since the bottleneck here was fetching data from redis that did not need to be fetched.

**Comment by kylebrandt** [comment]
should eventually fix #1662

**Comment by kylebrandt** [comment]
Also want to get this done before https://github.com/bosun-monitor/bosun/issues/1952 , as that will only increase the size of the rendered templates.

File: cmd/bosun/database/state_data.go [hunk 1]
```diff
@@ -51,16 +59,51 @@ type StateDataAccess interface {
 	GetLatestIncident(ak models.AlertKey) (*models.IncidentState, error)
 	GetAllOpenIncidents() ([]*models.IncidentState, error)
 	GetIncidentState(incidentId int64) (*models.IncidentState, error)
-	GetAllIncidents(ak models.AlertKey) ([]*models.IncidentState, error)
+	//GetAllIncidentIds() ([]int64, error)
```
Comment (kylebrandt) [inline cmd/bosun/database/state_data.go]: remove commented out func

File: cmd/bosun/database/state_data.go [hunk 2]
```diff
@@ -150,6 +193,45 @@ func (d *dataAccess) GetAllIncidents(ak models.AlertKey) ([]*models.IncidentStat
...
+		var err error
+		ids[i], err = strconv.ParseInt(strings.Split(sum, ":")[1], 0, 64)
+		if err != nil {
+			return nil, slog.Wrap(err)
+		}
+	}
+	return ids, nil
+
+}
+
```
Comment (kylebrandt) [inline cmd/bosun/database/state_data.go]: resolve descrepency between getAllIncidentIds and getAllIncideIdsByKeys ?

File: cmd/bosun/sched/check.go [hunk 3]
```diff
@@ -324,15 +336,16 @@ func (s *Schedule) executeTemplates(state *models.IncidentState, event *models.E
 			attachments = nil
 		}
 		state.Subject = string(subject)
-		state.Body = string(body)
+		//state.Body = string(body)
```
Comment (kylebrandt) [inline cmd/bosun/sched/check.go]: remove comment

File: cmd/bosun/database/migrate.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	LastAbnormalStatus models.Status
+	LastAbnormalTime   int64
+}
+
+func migrateRenderedTemplates(d *dataAccess) error {
+	slog.Infoln("Running rendered template migration")
+
+	// Hacky Work better?
+	ids, err := d.getAllIncidentIdsByKeys()
+	slog.Infoln("migrating %v incidents", len(ids))
```
Comment (kylebrandt) [inline cmd/bosun/database/migrate.go]: change to format func

Comment (kylebrandt) [inline cmd/bosun/database/migrate.go]: add comment to long entry that this can take several minutes.

File: cmd/bosun/sched/sched.go [hunk 5]
```diff
@@ -100,6 +100,9 @@ func (s *Schedule) Init(systemConf conf.SystemConfProvider, ruleConf conf.RuleCo
 	if s.Search == nil {
 		s.Search = search.NewSearch(s.DataAccess, skipLast)
 	}
+	if err := s.DataAccess.Migrate(); err != nil {
```
Comment (kylebrandt) [inline cmd/bosun/sched/sched.go]: this might be the wrong place to call this.

**Comment by kylebrandt** [comment]
seems incidents are disappearing after the migration... hrm...

**Comment by kylebrandt** [comment]
not good to have backing up Last redis data structure and migration going at same time I think.

**Comment by captncraig** [comment]
Looks like the 'allIncidents' key is only ever appended to. The count gets off when we purge/forget.

File: cmd/bosun/database/migrate.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+var tasks = []Migration{
+	{
+		UID:     "Migrate Rendered Templates",
+		Task:    migrateRenderedTemplates,
+		Version: 1,
+	},
+}
+
+type oldIncidentState struct {
+	Id       int64
```
Comment (captncraig) [inline cmd/bosun/database/migrate.go]:
```
type oldIncidentState struct{
 *models.IncidentState
 Body string 
 ...
}
```

Comment (captncraig) [inline cmd/bosun/database/migrate.go]: Then marshall `i.IncidentState`

File: cmd/bosun/database/state_data.go (hunk 2 above)
Comment (captncraig) [inline cmd/bosun/database/state_data.go]: Really we should just remove allIncidents altogether. It is not updated when you remove incidents. It does not look like we ever read from it. And you have shown we can get the incident data the hard way (with KEYS) if we need to.

File: cmd/bosun/sched/check.go [hunk 7]
```diff
@@ -125,8 +126,18 @@ func (s *Schedule) runHistory(r *RunHistory, ak models.AlertKey, event *models.E
...
+			if err != nil {
+				return
+			}
+			err = data.SetRenderedTemplates(incident.Id, rt)
+			if err != nil {
+				return
+			}
 		} else {
 			err = data.SetUnevaluated(ak, event.Unevaluated) // if nothing to save, at least store the unevaluated state
+			if err != nil {
```
Comment (captncraig) [inline cmd/bosun/sched/check.go]: This error checking block is a noop if there is no code following. Same thing directly above this block too.

File: cmd/bosun/sched/check.go [hunk 8]
```diff
@@ -183,7 +194,8 @@ func (s *Schedule) runHistory(r *RunHistory, ak models.AlertKey, event *models.E
 
 	//render templates and open alert key if abnormal
 	if event.Status > models.StNormal {
-		s.executeTemplates(incident, event, a, r)
+		s.executeTemplates(incident, rt, event, a, r)
+		s.DataAccess.State().SetRenderedTemplates(incident.Id, rt)
```
Comment (captncraig) [inline cmd/bosun/sched/check.go]: This happens in the defer as well. No need to set it twice.

File: cmd/bosun/sched/sched.go (hunk 5 above)
Comment (captncraig) [inline cmd/bosun/sched/sched.go]: Yes, I would do it from main if possible.

File: models/incidents.go [hunk 9]
```diff
@@ -22,11 +22,9 @@ type IncidentState struct {
...
 	Actions []Action `json:",omitempty"`
 
-	Subject      string
-	Body         string
-	EmailBody    []byte
-	EmailSubject []byte
-	Attachments  []*Attachment
+	Subject string
+
+	rt *RenderedTemplates
```
Comment (captncraig) [inline models/incidents.go]: remove this field

File: cmd/bosun/database/migrate.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+				version = SchemaVersion
+				return nil
+			}
+		} else {
+			return slog.Wrap(err)
+		}
+	}
+
+	for _, task := range tasks {
+		if task.Version > version { // Not sure logic is correct don't care right now.
```
Comment (kylebrandt) [inline cmd/bosun/database/migrate.go]: @captncraig Can you think through my logic for future versions and see if this is sane?

**Comment by kylebrandt** [comment]
Note: Migration for our (Stack Overflow) redis instance of 65436 incidents takes 2 min and 34 seconds on my workstation. Bosun is not accessible during the migration.

**Comment by kylebrandt** [comment]
Hrmm.. Still bad, seeing incidents with no body.

File: cmd/bosun/database/migrate.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+			// Check if migration has been run if not that run
+			err := task.Task(d)
+			if err != nil {
+				return slog.Wrap(err)
+			}
+		}
+		if _, err := conn.Do("SET", schemaKey, SchemaVersion); err != nil {
```
Comment (captncraig) [inline cmd/bosun/database/migrate.go]: should set to task.Version, not schemaVersion.

File: cmd/bosun/database/state_data.go [hunk 12]
```diff
@@ -13,6 +16,8 @@ import (
 /*
 incidentById:{id} - json encoded state. Authoritative source.
 
+renderedTemplatesById:{id} - json encoded RenderedTemplates ny Incident Id
```
Comment (captncraig) [inline cmd/bosun/database/state_data.go]: typo

File: cmd/bosun/database/state_data.go [hunk 13]
```diff
@@ -150,6 +192,45 @@ func (d *dataAccess) GetAllIncidents(ak models.AlertKey) ([]*models.IncidentStat
 	return d.incidentMultiGet(conn, ids)
 }
 
+func (d *dataAccess) getAllIncidentIds() ([]int64, error) {
```
Comment (captncraig) [inline cmd/bosun/database/state_data.go]: Why are these two functions both needed? Didn't you say you stopped using on or the other?

File: cmd/bosun/database/migrate.go (hunk 11 above)
Comment (kylebrandt) [inline cmd/bosun/database/migrate.go]: In which case I should set it in the loop after each task is successful then?

Comment (captncraig) [inline cmd/bosun/database/migrate.go]: yeah

**Pull request merged**